### PR TITLE
[io] Prefer lambdas over binds

### PR DESCRIPTION
**Pull request by SergioRAgostinho**

### CODE REVIEW

File: io/include/pcl/io/openni_camera/openni_device.h [hunk 1]
```diff
@@ -593,23 +593,23 @@ namespace openni_wrapper
   template<typename T> OpenNIDevice::CallbackHandle
   OpenNIDevice::registerImageCallback (void (T::*callback)(boost::shared_ptr<Image>, void* cookie), T& instance, void* custom_data) throw ()
   {
-    image_callback_[image_callback_handle_counter_] = boost::bind (callback, boost::ref (instance), _1, custom_data);
+    image_callback_[image_callback_handle_counter_] = [=, &instance] (boost::shared_ptr<Image> img) { (instance.*callback) (img, custom_data); };
```
Comment (SergioRAgostinho) [inline io/include/pcl/io/openni_camera/openni_device.h]: `Image` is an incomplete type in this header.

**Comment by SergioRAgostinho** [comment]
I'm scared about this PR and we have to be somewhat careful about it. The critical cases are situations in which I'm **capturing variables by reference**. Apparently bind was creating inner copies of these objects and there are now very little guarantees that some of these references are properly preserved during the lifetime of the lambda. Everything is running green at the moment but it might simply be untested behavior. I suspect we're gonna have a bunch of bug reports in the future which will be related to find the appropriate callback that required to have variables  capture as copy, instead of reference.

**Comment by taketwo** [comment]
What about using `std::bind` where we are not sure? I know we generally agreed to switch to lambdas, but I don't see why we need to be so dogmatic in the face of possible regressions.

**Comment by SergioRAgostinho** [comment]
That's a good idea. I'm sad it took me some time to reflect on this issue properly. I just hope the merges we already did don't come back to bite us.

**Comment by SergioRAgostinho** [comment]
I only preserved  lambdas with reference captures, in situations where the variables were being passed as `boost::ref(var)` to the bind. Also, I tried to expose the symbols in namespace `std::placeholders` through the entire file, but the boost placeholder symbols are still visible (somehow) and create a name clash. Hence my use of the full qualified name for the placeholders.

**Pull request merged**